src/context/blocked.rs

```diff
@@ -1,18 +1,151 @@
-use std::ptr;
+use crate::redismodule::{AUTH_HANDLED, AUTH_NOT_HANDLED};
+use crate::{raw, Context, ValkeyError, ValkeyString};
+use std::os::raw::{c_int, c_void};
 
-use crate::raw;
-use crate::Context;
+type AuthReplyCallback<T> =
+    fn(&Context, ValkeyString, ValkeyString, T) -> Result<c_int, ValkeyError>;
+type FreePrivDataCallback<T> = fn(&Context, T);
 
-pub struct BlockedClient {
+pub struct BlockedClient<T = ()> {
     pub(crate) inner: *mut raw::RedisModuleBlockedClient,
+    auth_reply_callback: Option<AuthReplyCallback<T>>,
+    free_priv_callback: Option<FreePrivDataCallback<T>>,
+    private_data: Option<Box<T>>,
+}
+
+#[allow(dead_code)]
+unsafe extern "C" fn auth_reply_wrapper<T: 'static + Clone>(
+    ctx: *mut raw::RedisModuleCtx,
+    username: *mut raw::RedisModuleString,
+    password: *mut raw::RedisModuleString,
+    err: *mut *mut raw::RedisModuleString,
+) -> c_int {
+    let context = Context::new(ctx);
+    let ctx_ptr = std::ptr::NonNull::new_unchecked(ctx);
+    let username = ValkeyString::new(Some(ctx_ptr), username);
+    let password = ValkeyString::new(Some(ctx_ptr), password);
+
+    // Get private data from context
+    let priv_data = context.get_blocked_client_privdata();
+    if priv_data.is_null() {
+        context.log_warning("[auth_reply_wrapper] Private data is null; this should not happen!");
+        return AUTH_NOT_HANDLED;
+    }
+
+    let callbacks = &*(priv_data
+        as *const (
+            AuthReplyCallback<T>,
+            Option<FreePrivDataCallback<T>>,
+            Box<T>,
+        ));
+
+    match (callbacks.0)(&context, username, password, callbacks.2.as_ref().clone()) {
+        Ok(status) => status,
+        Err(error) => {
+            let error_msg = ValkeyString::create(Some(ctx_ptr), error.to_string().as_str());
+            *err = error_msg.into_raw();
+            AUTH_HANDLED
+        }
+    }
+}
+
+#[allow(dead_code)]
+unsafe extern "C" fn free_callback_wrapper<T: 'static>(
+    ctx: *mut raw::RedisModuleCtx,
+    privdata: *mut c_void,
+) {
+    let context = Context::new(ctx);
+
+    if privdata.is_null() {
+        context
+            .log_warning("[free_callback_wrapper] Private data is null; this should not happen!");
+        return;
+    }
+
+    let (_, free_cb, data) = *Box::from_raw(
+        privdata
+            as *mut (
+                AuthReplyCallback<T>,
```

Comment:
Following the suggestion above (https://github.com/valkey-io/valkeymodule-rs/pull/186/files#r2019112792), I would recommend making this free_callback generic IF we are going with this approach